File: backend/src/main/java/moe/mewore/rabbit/geometry/Segment2D.java
package moe.mewore.rabbit.geometry;

public class Segment2D extends Line2D {

    private final Vector2 from;

    private final Vector2 to;

    private final double segmentMinConstant;

    private final double segmentMaxConstant;

    Segment2D(final Vector2 from, final Vector2 to) {
        super(from, to);
        this.from = from;
        this.to = to;

        final double fromPerpendicularConstant = xCoefficient * from.getY() - yCoefficient * from.getX();
        final double toPerpendicularConstant = xCoefficient * to.getY() - yCoefficient * to.getX();
        segmentMinConstant = Math.min(fromPerpendicularConstant, toPerpendicularConstant);
        segmentMaxConstant = Math.max(fromPerpendicularConstant, toPerpendicularConstant);
    }

    @Override
    public double distanceToPointSquared(final Vector2 point) {
        if (!pointFallsOnSegment(point)) {
            return Math.min(from.distanceToSquared(point), to.distanceToSquared(point));
        }
        return super.distanceToPointSquared(point);
    }

    private boolean pointFallsOnSegment(final Vector2 point) {
        final double pointConstant = xCoefficient * point.getY() - yCoefficient * point.getX();
        return pointConstant > segmentMinConstant && pointConstant < segmentMaxConstant;
    }
}

File: frontend/src/game/entities/geometry/segment-2d.ts
import { Line2D } from './line-2d';
import { Vector2Entity } from './vector2-entity';

export class Segment2D extends Line2D {
    private readonly segmentMinConstant: number;

    private readonly segmentMaxConstant: number;

    constructor(private readonly from: Vector2Entity, private readonly to: Vector2Entity) {
        super(from, to);

        const fromPerpendicularConstant = this.xCoefficient * from.y - this.yCoefficient * from.x;
        const toPerpendicularConstant = this.xCoefficient * to.y - this.yCoefficient * to.x;
        this.segmentMinConstant = Math.min(fromPerpendicularConstant, toPerpendicularConstant);
        this.segmentMaxConstant = Math.max(fromPerpendicularConstant, toPerpendicularConstant);
    }

    distanceToPointSquared(x: number, y: number): number {
        if (!this.pointFallsOnSegment(x, y)) {
            return Math.min(this.from.distanceToSquared(x, y), this.to.distanceToSquared(x, y));
        }
        return super.distanceToPointSquared(x, y);
    }

    pointFallsOnSegment(x: number, y: number): boolean {
        const pointConstant = this.xCoefficient * y - this.yCoefficient * x;
        return pointConstant > this.segmentMinConstant && pointConstant < this.segmentMaxConstant;
    }
}
